File: FrontEnd/lib/domain/repositories/i_notification_repository.dart
/// @file i_notification_repository.dart
/// @description Interfaz del repositorio de notificaciones in-app.
/// Define el contrato que los use cases usan para acceder a los datos.
/// La implementación concreta (NotificationRepositoryImpl) vive en data/.
/// @module Reminders
/// @layer Domain
library;

import '../entities/notification.dart';

/// Contrato del repositorio de notificaciones in-app.
///
/// Los use cases dependen de esta interfaz, nunca de la implementación concreta.
abstract interface class INotificationRepository {
  /// Devuelve las notificaciones del usuario autenticado, ordenadas por fecha descendente.
  ///
  /// [throws] AppError si hay error de red o el token es inválido.
  Future<List<AppNotification>> getUserNotifications();

  /// Marca notificaciones del usuario como leídas.
  /// Si se proporcionan [ids], solo marca esas; si no, marca todas.
  ///
  /// [throws] AppError si hay error de red.
  Future<void> markAllRead({List<String>? ids});

  /// Elimina notificaciones del usuario.
  /// Si se proporcionan [ids], solo elimina esas; si no, elimina todas.
  ///
  /// [throws] AppError si hay error de red.
  Future<void> deleteAll({List<String>? ids});
}

File: BackEnd/src/domain/repositories/INotificationRepository.ts
/**
 * @file INotificationRepository.ts
 * @description Interfaz del repositorio de notificaciones in-app.
 * @module Reminders
 * @layer Domain
 */

import type { Notification } from '../entities/Notification.js';

export interface INotificationRepository {
  /**
   * Devuelve las notificaciones de un usuario ordenadas por fecha descendente.
   *
   * @param userId — Id del usuario.
   * @returns Lista de notificaciones.
   */
  findByUserId(userId: string): Promise<Notification[]>;

  /**
   * Crea una notificación nueva.
   *
   * @param data — Datos sin id.
   * @returns Notificación creada.
   */
  create(data: Omit<Notification, 'id'>): Promise<Notification>;

  /**
   * Cuenta las notificaciones creadas hoy para un usuario.
   *
   * @param userId — Id del usuario.
   * @returns Número de notificaciones creadas hoy.
   */
  countTodayByUserId(userId: string): Promise<number>;

  /**
   * Marca como leídas todas las notificaciones de un usuario.
   *
   * @param userId — Id del usuario.
   */
  markAllReadByUserId(userId: string): Promise<void>;

  /**
   * Marca como leídas las notificaciones con los ids indicados (del usuario).
   *
   * @param userId — Id del usuario.
   * @param ids — Ids de notificaciones a marcar.
   */
  markReadByIds(userId: string, ids: string[]): Promise<void>;

  /**
   * Elimina todas las notificaciones de un usuario.
   *
   * @param userId — Id del usuario.
   */
  deleteAllByUserId(userId: string): Promise<void>;

  /**
   * Elimina las notificaciones con los ids indicados (del usuario).
   *
   * @param userId — Id del usuario.
   * @param ids — Ids de notificaciones a eliminar.
   */
  deleteByIds(userId: string, ids: string[]): Promise<void>;
}
